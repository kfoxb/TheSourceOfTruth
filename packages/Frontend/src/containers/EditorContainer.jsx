import React, { Component, Fragment } from 'react';
import firebase from 'firebase';
import PropTypes from 'prop-types';
import Editor from '../components/Editor';

export default class EditorContainer extends Component {
  static getCollection = (url) => {
    if (url.includes('journals')) {
      return 'journals';
    }
    return '';
  }

  static getDocumentId = (id) => {
    if (id === 'create') {
      return '';
    }
    return id;
  }

  static propTypes = {
    history: PropTypes.shape({
      replace: PropTypes.func.isRequired,
    }).isRequired,
    match: PropTypes.shape({
      url: PropTypes.string.isRequired,
      params: PropTypes.shape({
        id: PropTypes.string.isRequired,
      }).isRequired,
    }).isRequired,
  }

  constructor(props) {
    super(props);
    this.db = firebase.firestore();
    this.state = {
      collection: EditorContainer.getCollection(props.match.url),
      documentId: EditorContainer.getDocumentId(props.match.params.id),
      loading: true,
      title: '',
      value: '',
    };
  }

  componentDidMount() {
    if (!this.state.documentId) {
      this.db.collection(this.state.collection).add({
        title: '',
        value: '',
      })
        .then((docRef) => {
          this.props.history.replace(`/${this.state.collection}/${docRef.id}`);
          this.setState({
            documentId: docRef.id,
            loading: false,
          });
        })
        .catch((error) => {
          // eslint-disable-next-line no-console
          console.error('Could not create new document: ', error);
        });
    } else {
      this.db.collection(this.state.collection).doc(this.state.documentId).get()
        .then((doc) => {
          if (doc.exists) {
            const { title, value } = doc.data();
            this.setState({
              loading: false,
              title,
              value,
            }, () => {
              if (this.quill) {
                this.quill.focus();
                const editor = this.quill.getEditor();
                editor.setSelection(editor.getLength());
              }
            });
          } else {
            this.setState({ loading: false });
          }
        });
    }
  }

  onChangeSelection = (range, source, editor) => {
    if (!this.state.loading) {
      const getRange = () => {
        if (range === null) {
          return range;
        }
        const { index, length } = range;
        return { index, length };
      };

      this.setState({ value: editor.getHTML(), range: getRange() }, () => {
        this.updateDb({
          value: this.state.value,
          range: this.state.range,
        });
      });
    }
  }

  setRef = (ref) => { this.quill = ref; };

  updateDb = (value) => {
    this.db.collection(this.state.collection).doc(this.state.documentId).update(value)
    // eslint-disable-next-line no-console
      .catch(err => console.error('err writing value to firebase: ', err));
  }

  handleTitleChange = ({ target: { value } }) => {
    const update = { title: value };
    this.setState(update, () => {
      this.updateDb(update);
    });
  }

  render() {
    if (this.state.loading) {
      return (<p>Loading...</p>);
    }
    return (
      <Fragment>
        <input
          onChange={this.handleTitleChange}
          style={{ width: '100%' }}
          placeholder="Add New Title Here"
          value={this.state.title}
        />
        <Editor
          onChangeSelection={this.onChangeSelection}
          setRef={this.setRef}
          value={this.state.value}
        />
      </Fragment>
    );
  }
}
