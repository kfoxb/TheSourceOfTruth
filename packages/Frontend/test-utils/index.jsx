/* eslint-env jest */
/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { shallow } from 'enzyme';

export const createAssertWithPropsToMatchSnapshot = (Component, defaultProps) =>
  (extraProps) => {
    const props = { ...defaultProps, ...extraProps };
    const wrapper = shallow(<Component {...props} />);
    expect(wrapper).toMatchSnapshot();
  };
