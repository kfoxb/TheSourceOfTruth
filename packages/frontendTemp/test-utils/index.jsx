/* eslint-env jest */
/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { shallow } from 'enzyme';

export const makeShallowCreateWrapper = (Component, defaultProps) =>
  (extraProps) => {
    const props = { ...defaultProps, ...extraProps };
    return shallow(<Component {...props} />);
  };

export const createAssertWithPropsToMatchSnapshot = (Component, defaultProps) => {
  const createWrapper = makeShallowCreateWrapper(Component, defaultProps);
  return (extraProps) => {
    expect(createWrapper(extraProps)).toMatchSnapshot();
  };
};
