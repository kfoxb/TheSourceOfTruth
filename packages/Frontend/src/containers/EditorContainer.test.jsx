/* eslint-env jest */
import { createAssertWithPropsToMatchSnapshot, makeShallowCreateWrapper } from '../../test-utils';
import EditorContainer from './EditorContainer';

jest.mock('firebase');

describe('EditorContainer', () => {
  const defaultProps = {
    history: {
      replace: () => {},
    },
    match: {
      url: '',
      params: {
        id: '',
      },
    },
  };


  it('should render', () => {
    createAssertWithPropsToMatchSnapshot(EditorContainer, defaultProps)();
  });

  describe('journals', () => {
    const createWrapper = makeShallowCreateWrapper(EditorContainer, defaultProps);

    it('should set it\'s collection to journals when the url contains journals', () => {
      const extraProps = {
        match: {
          url: '/journals/edit/',
          params: {
            id: '',
          },
        },
      };

      const wrapper = createWrapper(extraProps);
      expect(wrapper.state().collection).toEqual('journals');
    });
  });
});
