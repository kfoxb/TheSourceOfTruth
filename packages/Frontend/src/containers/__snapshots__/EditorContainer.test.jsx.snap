// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorContainer should render 1`] = `
<p>
  Loading...
</p>
`;
